### PR TITLE
Add arm64 support

### PR DESCRIPTION
**Pull request by audacioustux**
solves https://github.com/grafana/k6-operator/issues/187

Pls make any changes if required.

### CODE REVIEW

File: .github/workflows/unit-test.yaml [hunk 1]
```diff
@@ -7,17 +7,17 @@ jobs:
   test:
     strategy:
       matrix:
-        go-version: [1.19.x]
-        k8s_version: [1.19.2, 1.24.1, 1.27.1]
+        go-version: [1.19]
+        k8s_version: [1.24.1, 1.27.1, 1.28.3]
```
Comment (yorugac) [inline .github/workflows/unit-test.yaml:11]: Please remove these changes.

File: .github/workflows/e2e-test.yaml [hunk 2]
```diff
@@ -9,23 +9,23 @@ jobs:
       - name: "Build:checkout"
         uses: actions/checkout@v2
       - name: "Build:buildx"
-        uses: docker/setup-buildx-action@v1
+        uses: docker/setup-buildx-action@v3
         with:
-          version: v0.9.1 # Buildx version
```
[the remark this answers is not in the document]
Comment (audacioustux) [reply .github/workflows/e2e-test.yaml:14]: I see... the version bump worked for the push.yml workflow, so didn't check if it works for the e2e, my bad. Will check whether the e2e works completely fine.

**Comment by audacioustux** [comment]
@yorugac thanks for the feedback, I'll send individual PRs as you mentioned.
Yes, have checked the ARM Images, and it works https://github.com/audacioustux/k6-operator/pkgs/container/k6-operator

**Comment by yorugac** [comment]
Hi @audacioustux, it's been awhile; hope you're well :slightly_smiling_face: 
How is it going? Would you be able to continue with this PR?

**Comment by audacioustux** [comment]
Hi @yorugac, sorry got busy with other stuffs. https://github.com/audacioustux/k6-operator I could make the ARM64 support work here, but that was a bit too involved than I had anticipated. + The fact that there's hard-coded image uri in the code made it a bit difficult to make a well-tested PR. 
Also, I couldn't make the browser tests work https://github.com/grafana/xk6-browser/issues/1239 (I think it's not related to Arm though) and couldn't manage the time to look into it either.

I think the first thing that needs to be done, is to fix that hardcoded container image URI issue (for example: https://github.com/grafana/k6-operator/blob/c2e81d8c83297561e4539a2377dbd90ad9dab887/pkg/resources/jobs/stopper.go#L19) 

For the ARM64 support, I can hopefully send the PR in next weekend (30/31 March), sorry for keep you waiting.

**Comment by yorugac** [comment]
Hi @audacioustux, thanks for quick response!

Yes, the browser tests are a "known" issue: #289 AFAIR, some cases have worked and some didn't. Investigating this wasn't prioritized yet but good thing you've opened the issue in xk6-browser repo as well :+1: 

> I think the first thing that needs to be done, is to fix that hardcoded container image URI issue

:thinking: you meant for testing? To test custom images, you can pass them as part of `TestRun` CR definition, i.e. `starter.image` in this example.

> For the ARM64 support, I can hopefully send the PR in next weekend (30/31 March)

That would be amazing; thank you.

Btw, re-reading this PR: I've asked to have CI updates as separate PRs -- would you like me to take those, to reduce the "change request"?

**Comment by audacioustux** [comment]
> you meant for testing? To test custom images, you can pass them as part of TestRun CR definition, i.e. starter.image in this example.

Umm, as far I can remember, even though starter and runner image was declared in TestRun manifest, it was spawning grafana/k6-operator. May have to re-check that actually...

> Btw, re-reading this PR: I've asked to have CI updates as separate PRs -- would you like me to take those, to reduce the "change request"?

Yes please.

**Comment by audacioustux** [comment]
@yorugac Hi, pls let me know if the push works correctly and if anything needs to be changed.

File: Dockerfile.starter [hunk 3]
```diff
@@ -1,5 +1,3 @@
-FROM radial/busyboxplus:curl
+FROM curlimages/curl:latest
 
-# as defined by distroless for nonroot
-USER 65532:65532
-COPY --from=redboxoss/scuttle:latest /scuttle /bin/scuttle
+COPY --from=kvij/scuttle:latest /scuttle /bin/scuttle
```
Comment (yorugac) [inline Dockerfile.starter:3]: `nonroot` user definitely shouldn't be removed here. 
On that note, are these two Docker changes necessary for ARM support? If they aren't, again, I'd suggest moving them out into a separate PR.

**Review by yorugac — CHANGES_REQUESTED** [review]
Hi @audacioustux, Thanks for the update!

On CI changes in `push.yaml`: I've gone through these ones by now and they seem to be mostly OK.
`buildx` question remains: it seems like the issue we encountered was around ~ `v0.10` so maybe `v0.12` will be fine. If this version change is not necessary for ARM, I'd prefer to see it in another PR.

Secondly, on Dockerfiles: there is a change request there.

> let me know if the image build and push works correctly

To test this workflow (without further changes), I'll need to merge this PR first :joy: In general, that's why I request as minimal version of changes as possible: it can be pretty time-consuming to debug CI issues so it's best when they're small and atomic.

File: Dockerfile.starter (hunk 3 above)
Comment (audacioustux) [reply Dockerfile.starter:3]: yes, those are required for ARM64 support, as the current ones are old/unmaintained and lacks arm64 support. is the USER really necessary for curlimages/curl:latest(https://github.com/curl/curl-container)?

**Comment by audacioustux** [comment]
> On CI changes in `push.yaml`: I've gone through these ones by now and they seem to be mostly OK. `buildx` question remains: it seems like the issue we encountered was around ~ `v0.10` so maybe `v0.12` will be fine. If this version change is not necessary for ARM, I'd prefer to see it in another PR.

umm which issue exactly? I updated the version just to be on safe side and hoped it's most likely backward compatible. I wasn't sure if the "platform" value is supported in the previous version, so tried to save one trial and error.

> To test this workflow (without further changes), I'll need to merge this PR first 😂 In general, that's why I request as minimal version of changes as possible: it can be pretty time-consuming to debug CI issues so it's best when they're small and atomic.

I see. that's actually the issue I was referring to earlier. I think the current situation of the repo is a bit hard to rollout new changes or for community contributions. I actually had plan for adding Devcontainers support, that I think should significantly make things easier and streamlined - we may discuss about that later at some point.

File: Dockerfile.starter (hunk 3 above)
Comment (yorugac) [reply Dockerfile.starter:3]: >  those are required for ARM64 support, as the current ones are old/unmaintained and lacks arm64 support
OK, thanks for clarification.

> is the USER really necessary for curlimages/curl:latest(https://github.com/curl/curl-container)?
Yes. It was added deliberately. Currently it is present in both controller and starter; please keep those.

**Comment by yorugac** [comment]
On `buildx`: yes, I meant that issue. Basically Docker manifests got broken with newer version. I've been checking the GH changelogs: it seems like it's _still_ an issue even now... There might be a way to make a fix without pinning a version but for now let's not change that line with `buildx` version. I think ARM build should work anyway, right?

File: .github/workflows/push.yaml [hunk 4]
```diff
@@ -28,38 +28,43 @@ jobs:
...
         run: |
           echo "IMAGETAG=${{env.IMAGETAG}}"
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
       - name: "Build:buildx"
-        uses: docker/setup-buildx-action@v1
+        uses: docker/setup-buildx-action@v3
         with:
-          version: v0.9.1 # Buildx version
+          version: v0.12.1 # Buildx version
```
Comment (yorugac) [inline .github/workflows/push.yaml:36]: IOW, without this bump ^

File: Dockerfile.starter (hunk 3 above)
Comment (audacioustux) [reply Dockerfile.starter:5]: >  Currently it is present in both controller and starter; please keep those.

added USER 65532:65532 now

File: .github/workflows/e2e-test.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (audacioustux) [reply .github/workflows/e2e-test.yaml]: buildx version reverted back to 0.9.1 now

**Comment by audacioustux** [comment]
https://github.com/audacioustux/k6-operator/actions/runs/8500731873/job/23282902694 the push.yml worked with the buildx version pinned to 0.9.1, kept the `USER 65532:65532` in starter dockerfile... tested in k8s and working

**Review by yorugac — APPROVED** [review]
Great, thanks @audacioustux! :raised_hands:

**Comment by Israphel** [comment]
which version has ARM support?
